### PR TITLE
Insert refactor; handle node resizing on update

### PR DESCRIPTION
**Pull request by danlaine**

### CODE REVIEW

File: firewood/src/merkle.rs [hunk 1]
```diff
@@ -473,385 +440,356 @@ impl<T: WriteLinearStore> Merkle<T> {
...
+                };
+                // `last_node` is at a strict prefix of `key`.
+                // `remaining_key` is `key` after removing the prefix of `last_node` and 1 additional
+                // nibble for the child index that points to the node after `last_node`.
+                let remaining_key = &remaining_key[1..];
+
+                // See if the `last_node` has a child where `key` would go.
+                let child_option = *last_node_branch
+                    .children
+                    .get(last_node_to_child_index as usize)
```
Comment (rkuris) [inline firewood/src/merkle.rs:522]: `child_option` is not an option.

nit: we should think about better helper methods in `BranchNode` so we can just do: `last_node_branch.child(last_node_to_child_index)`. If we do this, then the `mut_child` should be `child_mut`.

File: firewood/src/merkle.rs [hunk 2]
```diff
@@ -473,385 +440,356 @@ impl<T: WriteLinearStore> Merkle<T> {
...
-                                // There is no parent of this leaf. It must be the root.
-                                self.set_root(new_addr)?;
-                                return Ok(hash_invalidation_addresses);
-                            };
+                        partial_path: Path::from_nibbles_iterator(remaining_key.iter().copied()),
+                    }))?;
+
+                    // Add new leaf as child of `last_node`.
+                    let mut updated_last_node_children = last_node_branch.children;
+                    *updated_last_node_children
```
Comment (rkuris) [inline firewood/src/merkle.rs:540]: `use last_node_branch.mut_child`

File: firewood/src/merkle.rs [hunk 3]
```diff
@@ -473,385 +440,356 @@ impl<T: WriteLinearStore> Merkle<T> {
...
+                                    value: Some(value),
+                                    children: new_branch_children,
+                                    child_hashes: Default::default(),
+                                })))?;
+
+                            let mut last_node_children = last_node_branch.children;
+                            let mut last_node_children_hashes =
+                                last_node_branch.child_hashes.clone();
+
+                            *last_node_children
```
Comment (rkuris) [inline firewood/src/merkle.rs:613]: `mut_child` here too probably

File: firewood/src/merkle.rs [hunk 4]
```diff
@@ -473,385 +440,356 @@ impl<T: WriteLinearStore> Merkle<T> {
...
-                                    value: child_branch.value.clone(),
-                                    child_hashes: child_branch.child_hashes.clone(),
-                                };
-                                let updated_child_branch =
-                                    Node::Branch(Box::new(updated_child_branch));
-                                update_always_shrinks!(
-                                    self.update_node(child_addr, updated_child_branch)
-                                )?;
+                        *new_last_node.mut_child(last_node_to_child_index) = Some(new_branch_addr);
+                        *new_last_node
```
Comment (rkuris) [inline firewood/src/merkle.rs:684]: mut_child

File: firewood/src/merkle.rs [hunk 5]
```diff
@@ -1356,6 +1294,9 @@ mod tests {
 
         merkle.insert(&key, Box::new(val)).unwrap();
         merkle.insert(&key_2, Box::new(val_2)).unwrap();
+        let got = merkle.get(&key_2).unwrap().unwrap(); // todo remove
```
Comment (rkuris) [inline firewood/src/merkle.rs:1297]: cleanup

File: firewood/src/merkle.rs [hunk 6]
```diff
@@ -1471,7 +1412,7 @@ mod tests {
         let mut merkle = Merkle::new(HashedNodeStore::initialize(MemStore::new(vec![])).unwrap());
         for (k, v) in items.iter() {
             merkle.insert(k.as_ref(), Box::from(v.as_ref()))?;
-            println!("{}", merkle.dump()?);
+            // println!("{}", merkle.dump()?);
```
Comment (rkuris) [inline firewood/src/merkle.rs:1415]: Interesting question whether we want this or not

We can consider moving it to debug logging, but it was super handy during insert testing!

Comment (danlaine) [reply firewood/src/merkle.rs]: uncommented. Let's leave it for now

File: firewood/src/merkle.rs (hunk 1 above)
Comment (danlaine) [reply firewood/src/merkle.rs]: Added `child` method and renamed `mut_child` to `child_mut`

File: firewood/src/merkle.rs (hunk 2 above)
Comment (danlaine) [reply firewood/src/merkle.rs:531]: Not sure I follow. We're immutably borrowing `last_node_branch` so I don't think we can modify a child with `child_mut` here

**Review by rkuris — APPROVED** [review]
reviewed in person

**Pull request merged**